### PR TITLE
One bridge to rule them all

### PR DESCRIPTION
**Pull request by titus6012000**
One bridge to handle all type of file.
Just need to add element in the configuration to handle new type of file

### CODE REVIEW

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java [hunk 1]
```diff
@@ -6,97 +6,122 @@
...
+                                .remoteDirectory(bridge.getFtpDirectory())
+                                .filter(fileListFilter)
+                                .autoCreateLocalDirectory(true)
+                                .localFilter(new FileSystemPersistentAcceptOnceFileListFilter(new SimpleMetadataStore(), ""))
+                                .localDirectory(Files.createTempDirectory(SYNCHRONIZE_TEMP_DIRECTORY_PREFIX).toFile()),
+                    e -> e.id(bridge.getBridgeIdentifiant() + "_sftp")
+                                .autoStartup(true)
+                                .poller(Pollers.fixedDelay(polling))
+
+                )
```
Comment (TheMaskedTurtle) [inline src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java:110]: Can be refactored a bit to be more readable

**Review by TheMaskedTurtle — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 2]
```diff
@@ -88,5 +88,13 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.24</version>
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply pom.xml:95]: variable created

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java [hunk 3]
```diff
@@ -30,36 +30,42 @@
 public class FileTransferFlow {
 
     private static final SpelExpressionParser PARSER = new SpelExpressionParser();
+    private final ApplicationContext applicationContext;
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java:33]: it is used 4 time. 
But when you replace ApplicationContext by AutowireCapableBeanFactory, indeeed it is unsed

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java [hunk 4]
```diff
@@ -30,36 +30,42 @@
...
 
-    @Value("${data-bridge.file-regex}")
-    private String fileNameRegex;
+    private RemoteFileConfiguration remoteFilesConfiguration;
 
-    @Bean
-    public MessageChannel archivesChannel() {
-        return new DirectChannel();
+    public FileTransferFlow(ApplicationContext applicationContext, RemoteFileConfiguration remoteFilesConfiguration) {
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java:37]: replaced

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java [hunk 5]
```diff
@@ -30,36 +30,42 @@
 public class FileTransferFlow {
 
     private static final SpelExpressionParser PARSER = new SpelExpressionParser();
+    private final ApplicationContext applicationContext;
 
-    @Value("${data-bridge.file-regex}")
-    private String fileNameRegex;
+    private RemoteFileConfiguration remoteFilesConfiguration;
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java:35]: Yes it is used only in the constructor

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java [hunk 6]
```diff
@@ -30,36 +30,42 @@
...
-    public MessageChannel archivesChannel() {
-        return new DirectChannel();
+    public FileTransferFlow(ApplicationContext applicationContext, RemoteFileConfiguration remoteFilesConfiguration) {
+        this.remoteFilesConfiguration = remoteFilesConfiguration;
+        this.applicationContext = applicationContext;
+        this.remoteFilesConfiguration.getDataBridgeList().stream().forEach(bridge -> {
+            MessageChannel archivesChannel = new DirectChannel();
+            this.applicationContext.getAutowireCapableBeanFactory().initializeBean(archivesChannel, bridge.getBridgeIdentifiant() + "_archives_channel");
+            MessageChannel filesChannel = new DirectChannel();
+            this.applicationContext.getAutowireCapableBeanFactory().initializeBean(filesChannel, bridge.getBridgeIdentifiant() + "_files_channel");
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java:44]: the 2 DirectChannel has been removed

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java [hunk 7]
```diff
@@ -76,8 +82,8 @@ private Message<File> addFileNameHeader(Message<File> message) {
                 .build();
     }
 
-    private boolean isFormatOk(String filename) {
-        return filename.matches(fileNameRegex);
+    private boolean isFormatOk(String filename, String regex) {
+        return filename.matches(regex);
     }
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/FileTransferFlow.java:88]: removed the empty line after the method

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/MinioSink.java [hunk 8]
```diff
@@ -14,23 +14,32 @@
...
 
+import javax.annotation.PostConstruct;
+
 /**
  * @author Amira Kahya {@literal <amira.kahya at rte-france.com>}
  */
 @Configuration
 public class MinioSink {
     public static final String AWS_CLIENT_SIGNER_TYPE = "AWSS3V4SignerType";
+    private final ApplicationContext applicationContext;
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/MinioSink.java:37]: AutowireCapableBeanFactory used

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java [hunk 9]
```diff
@@ -38,32 +38,27 @@
  * @author Amira Kahya {@literal <amira.kahya at rte-france.com>}
  */
 @Configuration
-@ConditionalOnProperty(prefix = "data-bridge.sources.ftp", name = "active", havingValue = "true")
+@ConditionalOnProperty(prefix = "data-bridges.ftp", name = "active", havingValue = "true")
 public class FtpSource {
     public static final String SYNCHRONIZE_TEMP_DIRECTORY_PREFIX = "gridcapa-data-bridge";
 
     private final ApplicationContext applicationContext;
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java:45]: AutowireCapableBeanFactory used

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java [hunk 10]
```diff
@@ -76,34 +71,58 @@ private SessionFactory<FTPFile> ftpSessionFactory() {
...
-        fileListFilter.addFilter(new FtpRegexPatternFileListFilter(String.join("|", remoteFileConfiguration.getRemoteFileRegex())));
-        fileSynchronizer.setFilter(fileListFilter);
-        return fileSynchronizer;
-    }
+    @PostConstruct
+    public void allFTP() throws IOException {
+        for (DataBridge bridge : remoteFilesConfiguration.getDataBridgeList()) {
+
+            MessageChannel channel = new PublishSubscribeChannel();
+            this.applicationContext.getAutowireCapableBeanFactory().initializeBean(channel, bridge.getBridgeIdentifiant() + "_ftp");
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java:79]: deleted

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java [hunk 11]
```diff
@@ -76,34 +71,58 @@ private SessionFactory<FTPFile> ftpSessionFactory() {
...
+
+            IntegrationFlow flow = IntegrationFlows.from(bridge.getBridgeIdentifiant() + "_flow")
+                    .channel("archivesChannel")
+                    .get();
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java:83]: removed

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java [hunk 12]
```diff
@@ -76,34 +71,58 @@ private SessionFactory<FTPFile> ftpSessionFactory() {
...
-    @Bean
-    @InboundChannelAdapter(channel = "ftpSourceChannel", poller = @Poller(fixedDelay = "${data-bridge.sources.ftp.polling-delay-in-ms}"))
-    public MessageSource<File> ftpMessageSource() throws IOException {
-        FtpInboundFileSynchronizingMessageSource source =
-                new FtpInboundFileSynchronizingMessageSource(ftpInboundFileSynchronizer());
-        source.setLocalDirectory(Files.createTempDirectory(SYNCHRONIZE_TEMP_DIRECTORY_PREFIX).toFile());
-        source.setAutoCreateLocalDirectory(true);
-        source.setLocalFilter(new FileSystemPersistentAcceptOnceFileListFilter(new SimpleMetadataStore(), ""));
-        return source;
+            this.applicationContext.getAutowireCapableBeanFactory().initializeBean(flow, bridge.getBridgeIdentifiant() + "_flow");
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java:88]: removed

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java [hunk 13]
```diff
@@ -76,34 +71,58 @@ private SessionFactory<FTPFile> ftpSessionFactory() {
...
+                                .remoteDirectory(bridge.getFtpDirectory())
+                                .filter(fileListFilter)
+                                .autoCreateLocalDirectory(true)
+                                .localFilter(new FileSystemPersistentAcceptOnceFileListFilter(new SimpleMetadataStore(), ""))
+                                .localDirectory(Files.createTempDirectory(SYNCHRONIZE_TEMP_DIRECTORY_PREFIX).toFile()),
+                    e -> e.id(bridge.getBridgeIdentifiant() + "_ftp")
+                                .autoStartup(true)
+                                .poller(Pollers.fixedDelay(polling))
+
+                )
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java:110]: I defined the 2 parameters of the from in external variables. 
Each new variable are clearly separate and now we can see that the from has two parameters

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java [hunk 14]
```diff
@@ -6,97 +6,122 @@
...
  * @author Amira Kahya {@literal <amira.kahya at rte-france.com>}
  */
 @Configuration
-@ConditionalOnProperty(prefix = "data-bridge.sources.sftp", name = "active", havingValue = "true")
+@ConditionalOnProperty(prefix = "data-bridges.sftp", name = "active", havingValue = "true")
 public class SftpSource {
     public static final String SYNCHRONIZE_TEMP_DIRECTORY_PREFIX = "gridcapa-data-bridge";
 
-    private final RemoteFileConfiguration remoteFileConfiguration;
+    private final ApplicationContext applicationContext;
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java:43]: AutowireCapableBeanFactory used

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java [hunk 15]
```diff
@@ -6,97 +6,122 @@
...
+    public void allSFTP() throws IOException {
+        for (DataBridge bridge : remoteFilesConfiguration.getDataBridgeList()) {
+
+            MessageChannel channel = new PublishSubscribeChannel();
+            this.applicationContext.getAutowireCapableBeanFactory().initializeBean(channel, bridge.getBridgeIdentifiant() + "_sftp");
+
+            IntegrationFlow flow = IntegrationFlows.from(bridge.getBridgeIdentifiant() + "_flow")
+                    .channel("archivesChannel")
+                    .get();
+
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java:84]: removed

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java [hunk 16]
```diff
@@ -76,34 +71,58 @@ private SessionFactory<FTPFile> ftpSessionFactory() {
...
+
+            MessageChannel channel = new PublishSubscribeChannel();
+            this.applicationContext.getAutowireCapableBeanFactory().initializeBean(channel, bridge.getBridgeIdentifiant() + "_ftp");
+
+            IntegrationFlow flow = IntegrationFlows.from(bridge.getBridgeIdentifiant() + "_flow")
+                    .channel("archivesChannel")
+                    .get();
+
+            IntegrationFlow ms = this.ftpInboundFlow(bridge);
+            this.applicationContext.getAutowireCapableBeanFactory().initializeBean(ms, bridge.getBridgeIdentifiant() + "_adapter");
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/FtpSource.java:86]: renamed

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java [hunk 17]
```diff
@@ -6,97 +6,122 @@
...
 
-    @Bean
-    @InboundChannelAdapter(channel = "sftpSourceChannel", poller = @Poller(fixedDelay = "${data-bridge.sources.sftp.polling-delay-in-ms}"))
-    public MessageSource<File> sftpMessageSource() throws IOException {
-        SftpInboundFileSynchronizingMessageSource source = new SftpInboundFileSynchronizingMessageSource(sftpInboundFileSynchronizer());
-        source.setLocalDirectory(Files.createTempDirectory(SYNCHRONIZE_TEMP_DIRECTORY_PREFIX).toFile());
-        source.setAutoCreateLocalDirectory(true);
-        source.setLocalFilter(new FileSystemPersistentAcceptOnceFileListFilter(new SimpleMetadataStore(), ""));
-        return source;
+            this.applicationContext.getAutowireCapableBeanFactory().initializeBean(flow, bridge.getBridgeIdentifiant() + "_sftp_flow");
```
[the remark this answers is not in the document]
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java:88]: removed

File: src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java (hunk 1 above)
Comment (titus6012000) [reply src/main/java/com/farao_community/farao/gridcapa/data_bridge/sources/SftpSource.java:110]: refactor in the same way as FtpSource

**Comment by vbochetRTE** [comment]
Out of date. Replaced by https://github.com/farao-community/gridcapa-data-bridge/pull/43.

**Pull request closed without merging**